File: MonotonicStack/Easy/1475_Final_Prices_With_a_Special_Discount_in_a_Shop/Solution.cpp
#include <vector>
#include <stack>


class Solution {
public:
    std::vector<int> finalPrices(std::vector<int>& prices) {
        int n = prices.size();
        std::vector<int> res(prices);
        std::stack<int> st;

        for (int i{}; i < n; ++i) {
            while (!st.empty() && prices[i] <= prices[st.top()]) {
                res[st.top()] -= prices[i];
                st.pop();
            }

            st.push(i);
        }

        return res;
    }
};

File: MonotonicStack/Easy/1475_Final_Prices_With_a_Special_Discount_in_a_Shop/Solution.py
from typing import List


class Solution:
    def finalPrices(self, prices: List[int]) -> List[int]:
        res = prices[:]
        st = []

        for i, price in enumerate(prices):
            while st and prices[i] <= prices[st[-1]]:
                res[st[-1]] -= prices[i]
                st.pop()

            st.append(i)

        return res